pkg/appmain/appmain.go

```diff
@@ -28,27 +36,37 @@ var (
 	printVersion = flag.Bool("version", false, "Prints the version and exits")
 )
 
-func Run(name string, version string, logic task.TaskFn) {
+func Run(config Config, logic task.TaskFn) {
 	flag.Parse()
 
+	var err error
+
 	if *printVersion {
-		fmt.Fprintln(os.Stdout, version)
+		fmt.Fprintln(os.Stdout, config.Version)
 		os.Exit(ExitSuccess)
 	}
 
-	err := logger.Configure()
-	if err == nil {
-		logger.Info(name, ", v", version)
+	if err = sentry.Initialize(config.SentryConfig); err == nil {
+
+		err = logger.Configure()
+		if err == nil {
+			defer sentry.Close()
+			defer logger.Close()
+			logger.Info(config.Name, ", v", config.Version)
 
-		// Only available on Windows for cleaning up subprocesses
-		job := newJobObject()
-		defer job.Close()
+			// Only available on Windows for cleaning up subprocesses
+			job := newJobObject()
+			defer job.Close()
 
-		ctx, _ := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
+			ctx, _ := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
 
-		taskManager := task.NewTaskManager(ctx)
-		taskManager.GoFn("AppMain", logic)
-		err = taskManager.Wait()
+			taskManager := task.NewTaskManager(ctx)
+			taskManager.GoFn("AppMain", logic)
+			err = taskManager.Wait()
+			if err != nil {
+				logger.Error(err)
```

Comment:
Add os.Exit(ExitFailure) here. No reason to log the error here just for it to logged again down below. Instead, logging to sentry and then exiting would be good.

Reply:
os.Exit unfortunately doesn't call the deferred functions so anything through sentry/logger is lost if it hasn't been sent. I might adjust a few things so the error is returned from appmain.Run(...) and let the caller manage exiting etc

Reply:
Ah apologies @dave-juicelabs I thought you were referring to a os.Exit I had in there previously. I haven't actually made the change I said I was going to do just yet. But the logic here is that if the logger fails to init the error will be printed to stderr, while logging will goto stdout or a file.